corehq/apps/reports/models.py

```diff
@@ -375,29 +375,52 @@ def query_groups(self):
             raise TableauAPIError("Error: API does not work with more than 1000 groups on a single site.")
         return response_body['groups']['group']
 
-    def get_user_on_site(self, username):
+    def _make_paginated_request_for_users(self, method_name, additional_url_path):
+        page_size = 100
+        page_number = 1
+        total_users = sys.maxsize
+        tableau_users = []
+        print(additional_url_path)
+        while (page_size * (page_number - 1) < total_users):
+            print(page_number)
+            response_body = self._make_request(
+                self.GET,
+                method_name,
+                (self.base_url
+                + f'{additional_url_path}?pageSize={page_size}&pageNumber={page_number}'),
+                {}
+            )
+            # If it's the first page, grab the total user count.
+            if page_number == 1:
+                total_users = int(response_body['pagination']['totalAvailable'])
+                if total_users == 0:
+                    return []
+            tableau_users += response_body['users']['user']
```

Comment:
if you're interested, `yield from` is a nice way to do this sort of thing as a generator rather than building up a list, like
```suggestion
            yield from response_body['users']['user']
```

Reply:
No, not particularly - I'd say it's more personal preference than anything else.  I mentioned it because it removes a local state variable (`tableau_users`), so you could get rid of the initialization and `return` lines.

Another nice thing about using a generator when you don't need a list is memory usage.  Say you're fetching 10 pages of users, 1000 total.

* With a list, it'll stay in this function, making those 10 calls in a row, building up a list of 1000 users, where each element contains all the data returned from the API for that user.  Then it iterates through that list, extracting the username, ID, and site role, and building up a dict.
* With a generator, it will fetch one page of results, iterate through those, extracting the username, ID, and site role to start off the dict, then it'll fetch the second page, and so on.  This way it does everything in a single pass, and it never needs to store the full set of 1000 users until we get to the final data structure (the dict).

That generator type approach is common in functional programming, it's a nice way to separate the logical components of the code (it's nice to have the pagination stuff separate from the parsing stuff. as you've done here) without compromising on efficiency.  Again, I don't think it'd make a meaningful difference in this context, though.